### DISCUSSION TITLE
We want to hear from you - Here is a quick little poll about our latest releases

### DISCUSSION THREAD

**Discussion by ghiscoding** [6pac/SlickGrid#853] [Polls]
Just a simple poll to see what changes excited you the most from our last 3 releases. ❤️ 

Also if you are happy with our new version 5.0 and wish to showcase your project, feel free to do so in the discussion below by either posting a link to your project (preferably public) and/or add print screen below. 🚀 

Thank you

**Comment by syonfox** [comment]
Yay less dependencies boo more appendices ;) 

Also always like cool themes but this project is already large and stable don't break it to much haha

**Reply by ghiscoding** [reply]
I see you're in the camp of "if it ain't broken, don't fix it"... to which I would reply, "you can't make an omelet without breaking eggs"... you can always go back to v2.x with jQuery/jQueryUI if you call that stable ;). 

Also the new Alpine Theme is totally optional, you can still use the classic theme, just use what you want or create your own (nearly everything is optional in this lib and nearly all pieces are now extendable).

**Reply by PrimalNinja** [reply]
It would be great to get rid of typescript and if possible removes the requirement for jQueryUI - or at least makes it optional.

**Reply by ghiscoding** [reply]
> if possible removes the requirement for jQueryUI - or at least makes it optional.

You probably want to re-read the last 2 releases (Migrations Wikis), because jQueryUI is no longer required since v3 and we dropped jQuery in v4 (that's what the question form above is all about). You can still use jQuery/jQueryUI but it's now totally optional. The only new requirement is SortableJS which was introduced to replace jQueryUI.

https://github.com/6pac/SlickGrid/blob/36173fcc6ae621952f60f0d44a31f09c378990e7/package.json#L67-L69

> It would be great to get rid of typescript

most people would totally disagree with you, including myself, since most users would prefer to use projects that are written in TypeScript because it brings type safety and peace of mind that this safety is enforced. It also brings Type Definitions (`.d.ts`) that many users are interested in since it brings intellisense in code editors for all available options (there are many in SlickGrid)... It took a lot of effort to migrate to TypeScript and we have no intention to rolling it back to plain JS, note however that the 3 new builds will always be transpiled to plain JS so the build size differences are minimal. The ES6/ESM builds would also not be possible without TypeScript, at least not with a lot more effort.

I would suggest that you read all past Migrations to understand all changes applied, and as an alternative, you could also stick with v4 if you really want to stay with pure JS without TypeScript (though less recommended since we won't provide support for older versions)

**Comment by baj1210** [comment]
I'm also not a fan of the typescript.

"TypeScript is a [superset](https://en.wikipedia.org/wiki/Superset) of JavaScript, all JavaScript programs are [syntactically](https://en.wikipedia.org/wiki/Syntax_(programming_languages)) valid TypeScript, but they can fail to [type-check](https://en.wikipedia.org/wiki/Type_system#Type_checking) for [safety](https://en.wikipedia.org/wiki/Type_safety) reasons."

Since everything transpiles to javascript, and javascript is dynamically typed for a reason, I find it complicates the codebase and does not run error free out of the box. A javascript port would be ideal, I am having to port to js anyway, but I can understand why no one would want to this for the entire codebase considering how painful it is. 

Typescript is another language basically. 

@ghiscoding " you could also stick with v4 if you really want to stay with pure JS without TypeScript"

The V4 branch appears to also use typescript.

**Reply by ghiscoding** [reply]
v4 is for sure plain JavaScript since we switched to TypeScript in v5, you just have to go to the last [v4.x](https://github.com/6pac/SlickGrid/tree/4.1.7) tag to see this, so please first verify before saying something wrong. 

I seriously don't understand why you complain because at the end of the day JS code is produced by transpilation, and even the new [vite-demo](https://github.com/6pac/SlickGrid/tree/master/vite-demo) that I created is written in plain JavaScript. We use TypeScript to do the build because it brings Type safety and the types (`d.ts`) that goes with them which TypeScript users love. The only downside of TypeScript is that in some cases it produces a slightly larger transpiled JS build, **but** since EcmaScript is evolving and we can drop support for IE and transpile to newer ES2020-ES2021 target, then the transpiled output is almost a 1-to-1  in size which and won't grow the transpiled code anymore. 

You don't love TypeScript and that's fine, you have your own opinion but thinking that you represent the majority is **very far** from the reality. So please stop writing negative comments (I also just realized that both GitHub usernames below are yours) and instead please make comments that could be constructive and helpful to the projects and these complains are simply not useful. I also don't see the need to debate the JavaScript vs TypeScript, we will not switch back to JavaScript and that's the end of it.

**Comment by PrimalNinja** [comment]
Hi, In reverse engineering typescript slickgrid to JS i actually learned a lot on how flexible JS is. There are a lot of constructs i wouldn't have thought of using in JS. I often use similar constructs in assembly language.In any case, the codebase is very large, I really made a mess of the JS in my first attempt but I realise now I am better to progress the JS translation from typescript rather than to try make it typical JS code once more.I have had to make multiple enhancements to my own code base, mainly to do with focusing and navigating around.  Personally I think its a pitty that slickgrid Js was never further developed but I can also accept the TS version is quite evolved and satisfies the TS audience.When I have more time, I will make a JS attempt two which I will check in.  I did also have the starting of my own grid but it also is nowhere as advanced as slickgrid yet, not advanced enough for most projects.JulianSent on the go with Vodafone
-------- Original message --------From: baj1210 ***@***.***> Date: 30/10/24  8:12 pm  (GMT+10:00) To: 6pac/SlickGrid ***@***.***> Cc: Julian ***@***.***>, Comment ***@***.***> Subject: Re: [6pac/SlickGrid] We want to hear from you - Here is a quick little poll about our latest releases (Discussion #853) 
I'm also not a fan of the typescript.
"TypeScript is a superset of JavaScript, all JavaScript programs are syntactically valid TypeScript, but they can fail to type-check for safety reasons."
Since everything transpiles to javascript, and javascript is dynamically typed for a reason, I find it complicates the codebase and does not run error free out of the box. A javascript port would be ideal, I am having to port to js anyway, but I can understand why no one would want to this for the entire codebase considering how painful it is.
Typescript is another language basically.
@ghiscoding " you could also stick with v4 if you really want to stay with pure JS without TypeScript"
The V4 branch appears to also use typescript.

**Reply by baj1210** [reply]
"I really made a mess of the JS in my first attempt but I realise now I am better to progress the JS translation from typescript rather than to try make it typical JS code once more."

I am currently making a huge mess...I sympathize.

"Personally I think its a pitty that slickgrid Js was never further developed but I can also accept the TS version is quite evolved and satisfies the TS audience.When I have more time, I will make a JS attempt two which I will check in. "


That would be great, If I am able to find a method to port to JS without too puch pain, I'll post what I find. It is shame the JS slickgrid was abandoned since I like the flexibility of JS too and the fact that typescript does not seem to really add the protection in claims to which, to me, obviates it's use:

https://jonasnorlinder.github.io/Broken-world/

I think there may even be a way to compile the tsx files to js using tsconfig.json:

{
  "compilerOptions": {
    "target": "es6", // Adjust this to your desired ECMAScript version
    "module": "commonjs", // Change to "esnext" if you're using ES modules
    "jsx": "react-jsx", // Use "react-jsx" for the new JSX transform
    "allowJs": true, // Allows JavaScript files to be compiled
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "resolveJsonModule": true,
    "strict": true,
    "noImplicitAny": true,
    "skipLibCheck": true,
    "noEmit": false, // Ensure this is set to allow emitting output
    "outDir": "./dist" // Specify an output directory for compiled files
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules"]
}

**Comment by baj1210** [comment]
Ghislain,

Je vous prie de m'excuser pour toute critique implicite.I understand what
you are saying - complaints have no real use without suggestions.

"(I also just realized that @PrimalNinja <https://github.com/PrimalNinja>
and @baj1210 <https://github.com/baj1210> are both your accounts)" <----
this is not true. @PrimalNinja is not me, I would not do such a thing.

There is no negativity with my comments. I understand there is no interest
in porting the typescript back to javascript and I am trying to work out
that issue now. If I come up with a solution, I will contribute the code.

Overall, it is a fantastic project. The suggestion I have would be to
create a JS port so that it can be used with a variety of frameworks, but
you have already said there is no way this will ever happen, so what choice
do I have but to protest or complain. It is not personal or meant to
detract from your hard work.

" You don't love TypeScript and that's fine you have your own opinion but
thinking that you represent the majority is *very far* from the reality."
<--- Ghislain, I know you can understand how this is an assumption. I have
no idea how many people prefer typescript to javascript.

How many people complained that it was not in typescript before it was
ported to typescript from javascript? Then how many people have complained
about wanting it to be in javascript now that it is in typescript?

I think I might not even be using the correct branch?
https://github.com/6pac/SlickGrid/tree/4.1.7

I was referring to https://github.com/ghiscoding/slickgrid-react/tree/master

The typescript makes it difficult to integrate with react, but I am trying.

Think of it this way, the project is so good that I complain out of a
desire to be able to use it, not out of criticism.

**Reply by ghiscoding** [reply]
SlickGrid and Slickgrid-React are 2 separate projects that are completely separate for each other, Slickgrid-React doesn't even use or depend on `6pac/slickgrid` (it was using it a while ago but that's longer true since it uses Slickgrid-Universal which is now a standalone). I just happen to support all these SlickGrid projects.

**Comment by baj1210** [comment]
" I just happen to support all these SlickGrid projects."

Your efforts are certainly appreciated - it is a lot to manage.

**Comment by 6pac** [comment]
If I can just make a comment here. The Typescript in the latest SlickGrid build transpiles down to Javascript. This is the whole point of Typescript - it can't be run on anything as it is. The js target files are used in javascript projects, such as NodeJs or a web browser.  The pipeline produces a number of sets of target js files depending on the package management scheme (or not) that you use.

The only issue with TypeScript is that the transpiler can often produce quite ugly native js that is hard to follow and debug if you do so in the js world. The easiest solution to this is to set up the development tools correctly with TypeScript and allow the debugger to overlay the TS info into the debugging process.

I'm not a TypeScript user at the moment but I am in the early phases of converting my JS code to TS. @ghiscoding has already answered a few questions of mine in the forum re setting up the tools. I'll perhaps post to the wiki as I go, and any of you could too. Visual Studio seems to support it all quite well (I know that's what G uses), but I'm not sure of the support with VS Code or WebStorm which seem to be the two other most popular IDEs.

**Reply by baj1210** [reply]
My issue with typescript stems entirely from the fact that I am unfamiliar with it. I'll be open-minded towards learning it as I'm sure it has benefits.

I was having a difficult time with it due to my desire to integrate slickgrid into a react front-end that I have written in plain vanilla JS (even avoiding JSX).  

I did manage to end up porting the tsx files in **slickgrid-react-demos** to js by using a tsconfg.json file that looks like:

_tsconfig.json_ 
```

`{
  "compileOnSave": false,
  "compilerOptions": {
    "target": "es2022",
    "module": "esnext",
    "sourceMap": true,
    "allowSyntheticDefaultImports": true,
    "downlevelIteration": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "moduleResolution": "node",
    "allowJs": false,
    "baseUrl": "src",
    "isolatedModules": true,
    "noEmit": false, // Change to false to enable emitting JS files
    "resolveJsonModule": true,
    "skipLibCheck": true,
    "strict": true,
    "jsx": "react-jsx",
    "lib": [
      "dom", "dom.iterable", "ES2022"
    ],
    "types": ["node"],
    "typeRoots": [
      "node_modules/@types",
      "src/typings"
    ]
  },
  "exclude": [
    "react_project",
    "dist",
    "node_modules",
    "**/*.spec.ts"
  ],
  "filesGlob": [
    "./src/**/*.ts",
    "./src/**/*.tsx",
    "./test/**/*.ts",
    "./custom_typings/**/*.d.ts"
  ],
  "include": [
    "src"
  ]
}`
```


and also modifying the tsconfig.spec.json file as such:

_tsconfig.spec.json_

```
`
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist",
    "baseUrl": "./",
    "types": [
      "cypress",
      "node"
    ]
  },
  "include": [
    "**/*.spec.ts",
    "src/typings/**/*.ts",
    "**/*.d.ts"
  ],
  "typeAcquisition": { "include": [] }
}`
```


I then used webpack (specifically the **copy-webpack-plugin** to populate an output directory named /dist with the .js files which I could then bring into my current react codebase. 

_webpack.config.js_
```

`const path = require('path');
const CopyPlugin = require('copy-webpack-plugin');

module.exports = {
  entry: './src/index.js', // Change this to your entry file
  output: {
    path: path.resolve(__dirname, 'dist'), // Output directory
    filename: 'bundle.js', // Output file name
  },
  module: {
    rules: [
      {
        test: /\.js$/, // Include all .js files
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader',
          options: {
            presets: ['@babel/preset-env', '@babel/preset-react'],
          },
        },
      },
      {
        test: /\.scss$/, // Include all .scss files
        use: [
          'style-loader', // Inject CSS into the DOM
          'css-loader', // Turns CSS into CommonJS
          'sass-loader', // Compiles Sass to CSS
        ],
      },
      {
        test: /\.css$/, // Include all .css files
        use: ['style-loader', 'css-loader'], // Use these loaders
      },
    ],
  },
  resolve: {
    extensions: ['.js', '.jsx', '.scss'], // Resolve these file extensions
  },
  plugins: [
    new CopyPlugin({
      patterns: [
        {
          from: path.resolve(__dirname, 'src'), // Use the directory instead of glob
          to: path.resolve(__dirname, 'dist'), // Maintain structure
          globOptions: {
            dot: true,
            gitignore: true,
            ignore: ['*.js', '*.jsx', '*.scss'], // Exclude JS and SCSS from copying
          },
        },
      ],
    }),
  ],
  mode: 'development', // Change to 'production' for production build
};`

```

I would be happy to add these files to the repo as @ghiscoding showed me earlier, but figured I'd hold off doing so in the event this would not be a desired addition.

**Comment by 6pac** [comment]
Out of my league, but I'm also not a react user and have no idea about WebPack. Looks like you're using @ghiscoding 's slickgrid-react, as you should be. That's the main thing.

**Reply by baj1210** [reply]
I just touched JS for the first time 4 months ago so everything is out of my league atm.

**Reply by ghiscoding** [reply]
Sorry guys but this discussion is starting to be way off the principal subject of the Poll Survey above. If there's other topics to discuss then I think it should be in a separate discussion. Thank you
